src/coreclr/src/vm/mlinfo.cpp

```diff
@@ -2784,22 +2784,35 @@ MarshalInfo::MarshalInfo(Module* pModule,
                 // * Vector64<T>: Represents the __m64 ABI primitive which requires currently unimplemented handling
                 // * Vector128<T>: Represents the __m128 ABI primitive which requires currently unimplemented handling
                 // * Vector256<T>: Represents the __m256 ABI primitive which requires currently unimplemented handling
-                // * Vector<T>: Has a variable size (either __m128 or __m256) and isn't readily usable for inteorp scenarios
-
-                if (m_pMT->HasInstantiation() && (!m_pMT->IsBlittable()
-                    || m_pMT->HasSameTypeDefAs(g_pByReferenceClass)
-                    || m_pMT->HasSameTypeDefAs(g_pNullableClass)
-                    || m_pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__VECTOR64T))
-                    || m_pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__VECTOR128T))
-                    || m_pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__VECTOR256T))
+                // * Vector<T>: Has a variable size (either __m128 or __m256) and isn't readily usable for interop scenarios
+
+                if (IsFieldScenario())
+                {
+                    if (m_pMT->HasSameTypeDefAs(g_pByReferenceClass))
+                    {
+                        // We need to allow Nullable<T> and VectorX<T> fields for people who are using Marshal.SizeOf instead of Unsafe.SizeOf
+                        // for back-compat reasons, but we still don't want to allow ByReference<T>.
+                        m_resID = IDS_EE_BADMARSHAL_GENERICS_RESTRICTION;
```

Comment:
Was this disallowed in .NET Core 3.0?

Reply:
I will go check again to validate.

Reply:
Due to C# language restrictions around ref structs as generic type parameters, this was only possible by using the deprecated `Marshal.SizeOf(Type t)` signature (which is also marked `EditorBrowsableState.Never`), which has been deprecated since before the introduction of `ByReference<T>` and family. It is disallowed with the not-deprecated `Marshal.SizeOf<T>` signature.

I can still change this back if we feel that we should, but that would allow users to marshal a `Span<T>` as a structure, which I don't know if we want to allow (this block was explicitly put in place as a way to block marshalling `Span<T>`).

Reply:
I would keep all .NET Core 3.x -> 5.x potentially breaking changes for a separate issue / PR. When we mix bug fixes and breaking changes together, nobody can figure our what actually happened after the fact.